controllers/credential_controller_test.go

```diff
@@ -0,0 +1,274 @@
+
+	BeforeEach(func() {
+		// failed test runs that don't clean up leave resources behind.
+		keys := []string{credentialKeyName, privateCredentialKeyName}
+		for _, value := range keys {
+
+			cred := &civ1.Credential{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      value,
+					Namespace: "default",
+				},
+			}
+
+			_ = k8sClient.Delete(context.Background(), cred)
+		}
+	})
+
+	AfterEach(func() {
+		// Add any teardown steps that needs to be executed after each test
+		keys := []string{credentialKeyName, privateCredentialKeyName}
+		for _, value := range keys {
+
+			cred := &civ1.Credential{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      value,
+					Namespace: "default",
+				},
+			}
+
+			_ = k8sClient.Delete(context.Background(), cred)
+		}
+	})
+
+	Context("One crededential", func() {
+		It("Should handle credential correctly", func() {
+
+			spec := civ1.CredentialSpec{
+				Type:                 "container-registry",
+				WhitelistedPipelines: "github.com/estafette/.+",
+				AdditionalProperties: map[string]interface{}{
+					"repository": "estafette",
+					"private":    false,
+					"username":   "estafettesvc",
+					"password":   "supersecretpassword",
+				},
+			}
+
+			key := types.NamespacedName{
+				Name:      credentialKeyName,
+				Namespace: "default",
+			}
+
+			configMapKey := types.NamespacedName{
+				Name:      "estafette-external-credentials",
+				Namespace: "default",
+			}
+
+			toCreate := &civ1.Credential{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      key.Name,
+					Namespace: key.Namespace,
+				},
+				Spec:   spec,
+				Status: civ1.CredentialStatus{},
+			}
+
+			By("Creating the credential successfully")
+			Expect(k8sClient.Create(context.Background(), toCreate)).Should(Succeed())
+			time.Sleep(time.Second * 2)
+
+			fetched := &civ1.Credential{}
+			Eventually(func() string {
+				_ = k8sClient.Get(context.Background(), key, fetched)
+				return fetched.Status.ConfigMap
+			}, timeout, interval).Should(Equal(configMapKey.Name))
+
+			By("Creating configmap successfully")
+			fetchedConfigMap := &corev1.ConfigMap{}
+
+			expectedConfig := map[string]interface{}{
+				"name":                 credentialKeyName,
+				"type":                 "container-registry",
+				"whitelistedPipelines": "github.com/estafette/.+",
+				"repository":           "estafette",
+				"private":              false,
+				"username":             "estafettesvc",
+				"password":             "supersecretpassword",
+			}
+			expectedCredentialsData := map[string]interface{}{
+				"credentials": []interface{}{expectedConfig},
+			}
+			expectedYaml, err := yaml.Marshal(expectedCredentialsData)
+			Expect(err).To(BeNil())
+
+			Eventually(func() string {
+				_ = k8sClient.Get(context.Background(), configMapKey, fetchedConfigMap)
+				return fetchedConfigMap.Data["credentials-config.yaml"]
+			}, timeout, interval).Should(Equal(string(expectedYaml)))
+
+			By("Deleting the credential")
```

Comment:
It seems this single test is verifying multiple functional requirements. If possible please split them into separate more accurately described tests so we know which functional requirement is tested and if it breaks what's actually broken.